File: src/DBHandlerVB/DbProvider/MsSqlDataProvider.vb
Imports System.Data.SqlClient
Imports DBHandlerVB.DBHandlerVB.Interfaces

Namespace DBHandlerVB.DbProvider
    Public Class MsSqlDataProvider
        Implements IDBWrapper
        Private Property _ConnectionString As String
        Private _disposed As Boolean
        Private _dbConnection As IDbConnection
        Private _dbTransaction As IDbTransaction
        Private sqlcmd As SqlCommand

        Public ReadOnly Property CurrentConnectionState As ConnectionState Implements IDBWrapper.CurrentConnectionState
            Get
                If _dbConnection Is Nothing Then
                    Return ConnectionState.Closed
                End If
                Return _dbConnection.State
            End Get
        End Property

        Public Sub CloseConnection() Implements IDBWrapper.CloseConnection
            _dbConnection.Close()
            _dbConnection.Dispose()
        End Sub

        Public Sub New(ByVal connectionString As String)
            _ConnectionString = connectionString
            _dbConnection = Nothing
            _dbTransaction = Nothing
            sqlcmd = New SqlCommand()
        End Sub

        Public Function CreateConnection() As IDbConnection Implements IDBWrapper.CreateConnection
            If _dbConnection Is Nothing OrElse (CurrentConnectionState = ConnectionState.Broken OrElse CurrentConnectionState = ConnectionState.Closed) Then
                _dbConnection = New SqlConnection(_ConnectionString)
                _dbConnection.Open()
                Return _dbConnection
            End If

            Return _dbConnection
        End Function

        Public Function CreateCommand(ByVal commandText As String, ByVal commandType As CommandType, ByVal connection As IDbConnection) As IDbCommand Implements IDBWrapper.CreateCommand
            sqlcmd.CommandText = commandText
            sqlcmd.Connection = CType(connection, SqlConnection)
            sqlcmd.CommandType = commandType
            Return sqlcmd
        End Function

        Public Function CreateAdapter() As IDataAdapter Implements IDBWrapper.CreateAdapter
            Return New SqlDataAdapter(sqlcmd)
        End Function

        Public Function CreateParameter() As IDbDataParameter Implements IDBWrapper.CreateParameter
            Return sqlcmd.CreateParameter()
        End Function

        Public Function Begin() As IDbTransaction Implements IDBWrapper.Begin
            _dbTransaction = CreateConnection().BeginTransaction()
            sqlcmd.Transaction = CType(_dbTransaction, SqlTransaction)
            Return _dbTransaction
        End Function

        Public Sub Commit() Implements IDBWrapper.Commit
            If _dbTransaction IsNot Nothing Then
                _dbTransaction.Commit()
                sqlcmd.Parameters.Clear()
                _dbTransaction.Dispose()
            Else
                Throw New ApplicationException("Transaction Not Even Open.")
            End If
        End Sub

        Public Sub Rollback() Implements IDBWrapper.Rollback
            If _dbTransaction IsNot Nothing Then
                _dbTransaction.Rollback()
                sqlcmd.Parameters.Clear()
                _dbTransaction.Dispose()
            Else
                Throw New ApplicationException("Transaction Not Even Open.")
            End If
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            dispose(True)
            GC.SuppressFinalize(Me)
        End Sub

        Private Sub dispose(ByVal disposing As Boolean)
            If Not _disposed Then
                If disposing Then
                    If _dbConnection IsNot Nothing AndAlso CurrentConnectionState = ConnectionState.Open Then CloseConnection()
                    sqlcmd.Dispose()
                    _dbTransaction = Nothing
                    _dbConnection = Nothing
                    sqlcmd = Nothing
                End If
                _disposed = True
            End If
        End Sub

        Protected Overrides Sub Finalize()
            dispose(False)
        End Sub
    End Class
End Namespace

File: src/DBHandler/DbProvider/MsSqlDataProvider.cs
using DBHandler.Interfaces;
using System;
using System.Data;
using System.Data.SqlClient;

namespace DBHandler.DbProvider
{
    /// <summary>
    /// Author : Faris Jaafar
    /// Created Date : FEB 2019
    /// UPDATED : -
    /// VERSION : 1.0
    /// </summary>
    public sealed class MsSqlDataProvider : IDBWrapper
    {
        private string ConnectionString { get; set; }
        private bool _disposed;
        private IDbConnection _dbConnection;
        private IDbTransaction _dbTransaction;
        private SqlCommand sqlcmd;
        public ConnectionState CurrentConnectionState
        {
            get
            {
                return _dbConnection.State;
            }
        }
        public void CloseConnection()
        {
            _dbConnection.Close();
            _dbConnection.Dispose();
        }
        public MsSqlDataProvider(string connectionString)
        {
            ConnectionString = connectionString;
            _dbTransaction = null;
            _dbConnection = null;
            sqlcmd = new SqlCommand();
        }

        public IDbConnection CreateConnection()
        {
            //check db persistent ...
            if (_dbConnection == null || (CurrentConnectionState == ConnectionState.Broken
                || CurrentConnectionState == ConnectionState.Closed))
            {
                _dbConnection = new SqlConnection(ConnectionString);
                _dbConnection.Open();
                return _dbConnection;
            }
            return _dbConnection;
        }

        public IDbCommand CreateCommand(string commandText, CommandType commandType, IDbConnection connection)
        {
            sqlcmd.CommandText = commandText;
            sqlcmd.Connection = (SqlConnection)connection;
            sqlcmd.CommandType = commandType;
            return sqlcmd;
        }

        public IDataAdapter CreateAdapter()
        {
            return new SqlDataAdapter(sqlcmd);
        }

        public IDbDataParameter CreateParameter()
        {
            return sqlcmd.CreateParameter();
        }

        public IDbTransaction Begin()
        {
            _dbTransaction = CreateConnection().BeginTransaction();
            sqlcmd.Transaction = (SqlTransaction)_dbTransaction;
            return _dbTransaction;
        }

        public void Commit()
        {
            if(_dbTransaction != null)
            {
                _dbTransaction.Commit();
                sqlcmd.Parameters.Clear();
                _dbTransaction.Dispose();
            }else
                throw new ApplicationException("Transaction is not even open");
        }

        public void Rollback()
        {
            if(_dbTransaction != null)
            {
                _dbTransaction.Rollback();
                sqlcmd.Parameters.Clear();
                _dbTransaction.Dispose();
            }
            else
                throw new ApplicationException("Transaction is not even open");
        }

        #region "Dispose pattern"
        public void Dispose()
        {
            dispose(true);
            GC.SuppressFinalize(this);
        }
        private void dispose(bool disposing)
        {
            if (!_disposed)
            {
                if (disposing)
                {
                    if (_dbConnection != null && CurrentConnectionState == ConnectionState.Open)
                        CloseConnection();
                    sqlcmd.Dispose();

                    _dbTransaction = null;
                    _dbConnection = null;
                    sqlcmd = null;
                }
                _disposed = true;
            }
        }
        /// <summary>
        /// faris ,destructor
        /// </summary>
        ~MsSqlDataProvider()
        {
            dispose(false);
        }
        #endregion        
    }
}
